keps/sig-apps/3939-allow-replacement-when-fully-terminated/README.md

```diff
@@ -0,0 +1,877 @@
+
+This case was added due to a bug discovered with running IndexedJobs with Tensorflow.  
+See [Jobs create replacement Pods as soon as a Pod is marked for deletion](https://github.com/kubernetes/kubernetes/issues/115844) for more details.
+
+#### Story 2
+
+As a cloud user, users would want to guarantee that the number of pods that are running is exactly the amount that they specify.  
+Terminating pods do not relinguish resources so scarce compute resource are still scheduled to those pods.
+Replacement pods do not produce unnecessary scale ups.
+
+#### Story 3
+
+As a Job-level quota controller, I want to track the number of terminating pods,
+in addition to the active pods.
+
+See [Kueue: Account for terminating pods when doing preemption](https://github.com/kubernetes-sigs/kueue/issues/510) for an example of this.
+
+### Notes/Constraints/Caveats (Optional)
+
+#### The default job controller behavior
+
+Based on the [proposed API](#job-api-definition) below, the behavior of the
+job controller prior to this KEP is equivalent to
+`podReplacementPolicy: TerminatingOrFailed`.
+
+This behavior has the following semantic problems:
+- A terminating Pod might gracefully terminate as Succeeded, but it counts
+  towards `.status.failed` as soon as it's terminating.
+- When using podFailurePolicy, the controller might create a replacement Pod
+  before being able to evaluate the terminal state of the Pod. The replacement
+  Pod might be terminated due to the policy.
+
+In a Job v2 API, we should consider having the default behavior equivalent to
+`podReplacementPolicy: Failed`, given the above problems.
+We could even consider removing the proposed field `podReplacementPolicy`.
+
+But for backwards compatibility, in v1, we have to introduce a change of
+behavior as opt-in.
+
+#### When Pods enter a terminating state
+
+Pods can be marked for termination by several controllers, which we typically
+refer to as disruptions, such as: kubelet eviction, scheduler preemption, API eviction, etc.
+
+The job controller itself can delete running Pods, in the following scenarios:
+
+1. A job is over the `activeDeadlineSeconds`.  
+1. When the number of Pod failures reaches the `backoffLimit`.
+1. With `PodFailurePolicy` active and `FailJob` is set as the action.
+
+In all these situations, the Pod initially gets a `deletionTimestamp`
+and we interpret the pod as "terminating". Once the pod terminates, it gets
+a terminal `phase` (`Succeeded` or `Failed`).
+
+### Exponential Backoff for Pod Failures
+
+The job controller implements backoff delays to prevent fast recreation of
+continuously failing Pods.
+
+This behavior is internal (not configurable through the API) and it's orthogonal
+to this KEP. The behavior will be preserved as follows:
+- When `podReplacementPolicy: TerminatingOrFailed`, the backoff period counts from
+  the time the Pod is terminating or Failed.
+- When `podReplacementPolicy: Failed`, the backoff period counts from the time the
```

Comment:
When a pod is deleted it means a 10s gap before the previous pod reaches the `Failed` phase, and replacement is created. Wondering if this gap isn't too big?

Before, assuming 30s grace period for pod, there was 20s overlap of both the terminating pod and the replacement running. 

If we want to reduce the gap we could say that the first pod is recreated after 0s, or 1s, then next follow the expotential backoff delay. Alternatively, set the default backoff to 1s and grow expotentially, ie. 1s,2s.4s, 8s....
wdyt @alculquicondor @soltysh ?

Reply:
We could track this separately, but I wouldn't opposed to a reduced backoff.

Reply:
I seem to remember that there was an open KEP about configurable delay, but I can't find it.
Although it might be better to just offer more reasonable constants, considering the higher scalability of clusters compared to when the job controller was first written.

Reply:
This was an Issue, found it: https://github.com/kubernetes/kubernetes/issues/114651

Reply:
I'm fine with either lowered constants (but it is tricky to choose them), or configuration. Also, fine with tracking it independently. 

My feeling is that the 10s gap will not be good for some users. Sounds like something we may ask on batch wg meeting.